### PR TITLE
Add CBOR file format

### PR DESCRIPTION
**Pull request by ttser1085**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add CBOR input/output format

### CBOR format
Reading and writing binary data format CBOR(Concise Binary Object Representation) 

<table role="table">
<thead>
<tr>
<th>ClickHouse data type</th>
<th>CBOR</th>
</tr>
</thead>
<tbody>
<tr>
<td><code class="notranslate">UIntN, IntN</code></td>
<td><code class="notranslate">int or bignum</code></td>
</tr>
<tr>
<td><code class="notranslate">Enum8, Enum16</code></td>
<td><code class="notranslate">int</code></td>
</tr>
<tr>
<td><code class="notranslate">String, FixedString</code></td>
<td><code class="notranslate">string</code></td>
</tr>
<tr>
<td><code class="notranslate">Float32</code></td>
<td><code class="notranslate">float</code></td>
</tr>
<tr>
<tr>
<td><code class="notranslate">FLoat64</code></td>
<td><code class="notranslate">double</code></td>
</tr>
<tr>
<td><code class="notranslate">Date, Date32</code></td>
<td><code class="notranslate">date (tag 100)</code></td>
</tr>
<tr>
<td><code class="notranslate">DateTime, DateTime64</code></td>
<td><code class="notranslate">int</code></td>
</tr>
<tr>
<td><code class="notranslate">DecimalN</code></td>
<td><code class="notranslate">int or bignum</code></td>
</tr>
<tr>
<td><code class="notranslate">IPv4</code></td>
<td><code class="notranslate">int</code></td>
</tr>
<tr>
<td><code class="notranslate">IPv6</code></td>
<td><code class="notranslate">bytes</code></td>
</tr>
<tr>
<td><code class="notranslate">Array, Tuple</code></td>
<td><code class="notranslate">array</code></td>
</tr>
<tr>
<td><code class="notranslate">Map</code></td>
<td><code class="notranslate">map</code></td>
</tr>
<tr>
<td><code class="notranslate">Nullable, Nothing</code></td>
<td><code class="notranslate">null</code></td>
</tr>
<tr>
<td><code class="notranslate">UUID</code></td>
<td><code class="notranslate">string</code></td>
</tr>
</tbody>
</table>

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -363,6 +363,9 @@
 [submodule "contrib/idna"]
 	path = contrib/idna
 	url = https://github.com/ada-url/idna.git
+[submodule "contrib/cbor"]
```
Comment (Algunenano) [inline .gitmodules:366]: This code does not have a LICENSE and can't be used here.

File: tests/queries/0_stateless/02966_cbor_format.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+
+$CLICKHOUSE_CLIENT --query="DROP TABLE IF EXISTS cbor";
+
+$CLICKHOUSE_CLIENT --query="CREATE TABLE cbor (uint8 UInt8, uint16 UInt16, uint32 UInt32, uint64 UInt64, int8 Int8, int16 Int16, int32 Int32, int64 Int64, float Float32, double Float64, string String, date Date) ENGINE = Memory";
```
Comment (Algunenano) [inline tests/queries/0_stateless/02966_cbor_format.sh:10]: Consider using .sql tests instead. It will make things simpler and allow the fuzzer understand these queries and run more of them.

Comment (Algunenano) [reply tests/queries/0_stateless/02966_cbor_format.sh:10]: Also, it needs way more tests for both IN and OUT outputs

File: .gitmodules (hunk 1 above)
Comment (ttser1085) [reply .gitmodules:366]: LICENSE in the src folder

**Review by Algunenano — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 1 above)
Comment (Algunenano) [reply .gitmodules:366]: :+1:

**Review by Algunenano — COMMENTED** [review]
I'm not super confident about the chosen library (seems unmaintained and with poor code standards), but let's do a CI run and see.

Also, CLA needs to be approved by all commiters.

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 3ae81f0158f923a5f6c21781625d9391836e1cac with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62977/3ae81f0158f923a5f6c21781625d9391836e1cac/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
🕵 Checked the relevance of CBOR and the state of this PR. Unlike the sibling intern-task formats (`FlexBuffers` #63057, Amazon Ion #62961), CBOR is genuinely relevant — but this implementation is not the right vehicle for it.

**The format.** CBOR is an IETF standard ([RFC 8949](https://www.rfc-editor.org/rfc/rfc8949.html), STD 94) with an active working group and, more importantly, standards-mandated adoption that keeps growing:

- WebAuthn / FIDO2: CTAP2 messages and attestation objects are CBOR — every passkey registration and login on the web involves CBOR.
- COSE (RFC 9052) and CWT (RFC 8392) for signing and tokens.
- ISO/IEC 18013-5 mobile driver's licenses (mdoc) and the [EU Digital Identity Wallet](https://github.com/eu-digital-identity-wallet/eudi-doc-standards-and-technical-specifications/issues/82) are built on CBOR; the EU COVID certificates were CBOR/COSE.
- IoT: CoAP payloads, SenML (RFC 8428), LwM2M, Zephyr's `zcbor`.
- DAG-CBOR is the native encoding of IPFS/Filecoin.

Mature implementations exist in essentially every language. So on the "did it get traction" question, CBOR passes where Ion and FlexBuffers fail: it is effectively the IETF-standardized sibling of `MsgPack` — row-oriented, streaming-friendly (CBOR Sequences, RFC 8742), self-describing.

**The ClickHouse fit.** Two caveats. First, nobody asked: no issue has ever requested CBOR support; the PR comes from the intern task list (https://github.com/ClickHouse/ClickHouse/issues/58394). Second, most real-world CBOR is small single objects — credentials, attestations, tokens, sensor readings — not bulk tabular streams, and as a row format it overlaps almost entirely with the existing `MsgPack` and `BSONEachRow`. The most plausible ClickHouse touchpoint may actually be decoding CBOR *blobs stored in columns* (WebAuthn attestation data in security analytics, IoT payloads), which points to a `fromCBOR`-style function at least as much as to a file format.

**This implementation.** The vendored library `naphaso/cbor-cpp` is a hobby project: 29 commits total, last one on 2023-10-04, 81 stars, 7 open issues (Apache 2.0 via per-file headers, so licensing is fine). @Algunenano's 2024 review already flagged the library's quality and maintenance, questioned negative-integer correctness against the reference implementation, and asked for round-trip tests — none of which was addressed. The PR has been dormant since 2024 and is currently in conflict. If we want CBOR, it would be less work to restart on a maintained library (`tinycbor`, `libcbor`, `QCBOR`, or `jsoncons`) than to rehabilitate this branch.

**Bottom line.** The format is relevant and, of the three formats from that intern cohort, the only one with a live, growing ecosystem. But the concrete demand for it as a ClickHouse format is undemonstrated, and this PR's foundation (an unmaintained hobby library plus unaddressed correctness review) means the sensible options are: close it and keep CBOR as a tracked feature idea, or restart from scratch on a solid library if demand materializes.
